torch/csrc/distributed/rpc/ProcessGroupAgent.cpp

```diff
@@ -140,87 +133,111 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
     future->markCompleted();
   }
 
-  enqueue(SendWork(dstRank, std::move(message)));
+  enqueueSend(SendWork(dstRank, std::move(message)));
   return future;
 }
 
-void ProcessGroupAgent::enqueue(SendWork work) {
-  std::unique_lock<std::mutex> lock(sendQueueMutex_);
-  sendQueue_.emplace_back(std::move(work));
-  lock.unlock();
+void ProcessGroupAgent::enqueueSend(SendWork work) {
+  // NB: this can be changed to use a native move capture when moved to C++14
+  threadPool_.run(std::bind(
+    [&](const SendWork& work) {
+      std::stringstream ss;
+      serialize(work.message_, ss);
+      std::string serializedPayload = ss.str();
+
+      std::vector<torch::Tensor> preamble = {
+        torch::tensor(
+          {
+            (int64_t)pg_->getRank(),
+            (int64_t)serializedPayload.length(),
+            (int64_t)work.message_.type()
+          }, {torch::kLong})
+      };
+
+      // ProcessGroup is not thread-safe when sending with the same tag, hence
+      // the lock
+      std::vector<std::shared_ptr<c10d::ProcessGroup::Work>> pendingSends;
+      if (work.message_.isShutdown()) {
+        pendingSends.reserve(1);
+        std::lock_guard<std::mutex> guard(sendMutexes_[work.to_]);
+        pendingSends.emplace_back(
+            pg_->send(preamble, work.to_, work.to_ /* channelTag */));
+      } else {
+        std::vector<torch::Tensor> payload = {
+            torch::from_blob(
+                (void *)serializedPayload.c_str(),
+                serializedPayload.length(),
+                {torch::kChar}
+            )
+        };
+        pendingSends.reserve(2);
+        std::lock_guard<std::mutex> guard(sendMutexes_[work.to_]);
+        pendingSends.emplace_back(
+            pg_->send(preamble, work.to_, work.to_ /* channelTag */));
+        pendingSends.emplace_back(
+            pg_->send(payload, work.to_, work.to_ /* channelTag */));
+      }
+      for (auto& pendingSend: pendingSends) {
+        pendingSend->wait();
```

Comment:
curious about the reason you want to wait for gloo device thread?

Reply:
Because `work` will be destructed after the lambda function finishes, and the send function itself keep it alive. If destruction happens before send finish, the behavior will be undefined.

Reply:
@mrshenli 

Since the change you are making here, waiting for device thread to finish the work, is similar to, in network programming, waiting for os kernel to send out the buffer in kernel, making the user code essentially in a blocking fashion.

I wonder is this possible to create the worker object on heap `std::make_shared...` and capture the shared_ptr by value in lambda (adding ref count to the heap object) to keep it alive.

Reply:
@xush6528 could you please elaborate more? By `worker object` do you mean `SendWork`? If so, why using a `shared_ptr` would help? The `SendWork` is already capture by this lambda. Or are you referring to a different lambda than this one?

Reply:
Or we could have a separate GC thread for send, and enqueue all async send handles to that thread. The GC thread will then wait on all the send handle in order and delete it when done.

Reply:
@mrshenli  Sorry, I mean the `ProcessGroup::Work` you are waiting for. Updated the above comment.

Reply:
> Because work will be destructed after the lambda function finishes, and the send function itself does not keep the tensor alive. If destruction happens before send finish, the behavior will be undefined.

Oh, my bad, when I say `work` above, I actually mean `preamble` and `payload` tensors. We need keep those tensors alive before `send` finishes. And I agree that we can optimize this by capturing `ProcessGroup::Work` and those tensors (all as `std::shared_ptr`s) in a separate lambda, and wait there. Let me add an issue for this later. Thanks!

Reply:
See #24946